Synchronous notifications

manishtomar opened the pull request:
Provide an option to send notifications synchronously to endpoints. If enabled then any event generated as a result of processing a HTTP request is sent before the associated HTTP response is returned. If event sending fails the the HTTP request will also fail. This is useful if processing certain type of events are critical to functioning of the system. It is possible to lose events if there are many events pending in the memory queue and the registry instance crashes or restarts for some reason. This will also increase the latency of processing HTTP requests and hence should be used carefully for important events.

dmcgowan commented on notifications/sinks.go:
```diff
@@ -14,6 +14,51 @@ import (
+func (b *SyncBroadcaster) Write(events ...Event) error {
+	for _, sink := range b.sinks {
+		if err := sink.Write(events...); err != nil {
+			logrus.WithField("events", events).Errorf(
```
Can you use `.WithError`  here?

manishtomar replied:
This requires logrus update since current vendored version is old. Let me make another PR to update it.

dmcgowan commented on notifications/sinks.go:
```diff
@@ -14,6 +14,51 @@ import (
+		if err := sink.Write(events...); err != nil {
+			logrus.WithField("events", events).Errorf(
+				"syncbroadcaster: error writing events to %v, these events will be lost: %v", sink, err)
+			return err
```
What does returning an error here mean for a request? It seems this is aiming for consistency, but is it expected that a client would need to retry if 1 of n sinks failed to send?

manishtomar replied:
Returning error here results in failure of HTTP request (500). The intention is client will retry the request.

dmcgowan replied:
Retry but the original operation may have succeeded? Is there any case where the second request may go through a different code path because of the prior partial success?

manishtomar replied:
> Retry but the original operation may have succeeded?

Yes.

> Is there any case where the second request may go through a different code path because of the prior partial success?

Yes. In fact it happens currently where a blob push fails due to sink failure but the blob is actually stored. The client subsequently skips uploading since a `HEAD` check succeeds. This is an unfortunate flaw in the design.

caervs replied:
(e.g. engine pushes up a blob, blob gets written to storage backend but notifications fail, on second try engine does a HEAD request and sees blob is there?)

caervs commented on registry/handlers/app.go:
```diff
@@ -470,7 +471,9 @@ func (app *App) configureEvents(configuration *configuration.Configuration) {
 	// that the registry instances also act as the workers to keep deployment
 	// simple.
-	app.events.sink = notifications.NewBroadcaster(sinks...)
+	// NOTE(jmwong): Always use SyncBroadcaster here and delegate whether or not the events a processed asynchronously
```
@manishtomar 

a -> are -- also consider using your github alias since you're the commiter

caervs commented on notifications/sinks_test.go:
```diff
@@ -63,6 +63,75 @@ func TestBroadcaster(t *testing.T) {
+	}
+	// event was added to first sink but not second one
+	if len(good.events) != 2 {
+		t.Fatalf("sink should've 2 events")
```
should've -> should have (yay English for being consistent!)

GordonTheTurtle commented:
<!-- AUTOMATED:POULE:DCO-EXPLANATION -->
Please sign your commits following these rules:
https://github.com/moby/moby/blob/master/CONTRIBUTING.md#sign-your-work
The easiest way to do this is to amend the last commit:
~~~console
$ git clone -b "sync-notif" git@github.com:manishtomar/distribution.git somewhere
$ cd somewhere
$ git rebase -i HEAD~842353944800
editor opens
change each 'pick' to 'edit'
save the file and quit
$ git commit --amend -s --no-edit
$ git rebase --continue # and repeat the amend for each commit
$ git push -f
~~~

Amending updates the existing PR. You **DO NOT** need to open a new one.

manishtomar commented:
@dmcgowan I've updated logrus and a dependency package to use `WithError`. 
@caervs I've addressed nit comments. I've also added note in the doc about partial failure.

LMK if you've questions.

dmcgowan commented:
Overall this looks good, 2 concerns about error handling though. The first is that the eventQueue is going to be logging an additional warning after the error log after this change for non-sync cases. The second is related to what gets sent back to the client. I think this should be obvious from the error sent back to the client, ideally typed. If the request should be retried then a 503 may also be a better response code.

manishtomar commented:
I thought about this some more and realize that the failure scenario is not worth having this feature Closing. LMK if anyone has other opinions. Happy to talk through.
